cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -50,74 +50,71 @@ const (
 	kubeAPIServer         = "kube-apiserver"
 	kubeControllerManager = "kube-controller-manager"
 	kubeScheduler         = "kube-scheduler"
+
+	k8sCertsVolumeName   = "k8s-certs"
+	etcdVolumeName       = "etcd"
+	caCertsVolumeName    = "ca-certs"
+	caCertsVolumePath    = "/etc/ssl/certs"
+	caCertsPkiVolumeName = "ca-certs-etc-pki"
+	caCertsPkiVolumePath = "/etc/pki"
+	kubeConfigVolumeName = "kubeconfig"
 )
 
 // WriteStaticPodManifests builds manifest objects based on user provided configuration and then dumps it to disk
 // where kubelet will pick and schedule them.
 func WriteStaticPodManifests(cfg *kubeadmapi.MasterConfiguration) error {
-	volumes := []v1.Volume{k8sVolume()}
-	volumeMounts := []v1.VolumeMount{k8sVolumeMount()}
-
-	if isCertsVolumeMountNeeded() {
-		volumes = append(volumes, certsVolume(cfg))
-		volumeMounts = append(volumeMounts, certsVolumeMount())
-	}
-
-	if isPkiVolumeMountNeeded() {
-		volumes = append(volumes, pkiVolume())
-		volumeMounts = append(volumeMounts, pkiVolumeMount())
-	}
-
-	if !strings.HasPrefix(cfg.CertificatesDir, kubeadmapiext.DefaultCertificatesDir) {
-		volumes = append(volumes, newVolume("certdir", cfg.CertificatesDir))
-		volumeMounts = append(volumeMounts, newVolumeMount("certdir", cfg.CertificatesDir))
-	}
 
+	// TODO: Move the "pkg/util/version".Version object into the internal API instead of always parsing the string
 	k8sVersion, err := version.ParseSemantic(cfg.KubernetesVersion)
 	if err != nil {
 		return err
 	}
 
+	// Get the required hostpath mounts
+	volumes, volumeMounts := getHostPathVolumesForTheControlPlane(cfg)
+
 	// Prepare static pod specs
 	staticPodSpecs := map[string]v1.Pod{
 		kubeAPIServer: componentPod(v1.Container{
 			Name:          kubeAPIServer,
 			Image:         images.GetCoreImage(images.KubeAPIServerImage, cfg, kubeadmapi.GlobalEnvParams.HyperkubeImage),
-			Command:       getAPIServerCommand(cfg, false, k8sVersion),
-			VolumeMounts:  volumeMounts,
+			Command:       getAPIServerCommand(cfg, k8sVersion),
+			VolumeMounts:  volumeMounts[kubeAPIServer],
```

Comment:
The old schooler in me says you might want to hide this detail behind some utility functions, but not a concern at this time.